aten/src/ATen/core/autocast/register_autocast_ops.cpp

```diff
@@ -0,0 +1,528 @@
+  promote_with_tensorlist, // Separate from CastPolicy::promote due to lifetime concerns
+                           // of data viewed by TensorLists.
+};
+
+/********************************************************************
+Logic to extract the promote type from any Tensor or TensorList args.
+********************************************************************/
+
+// Overload to catch Tensor args.
+// If nextArg is floating-point, compare its scalar_type with our
+// current best guess for the promote type, and update if necessary.
+inline at::ScalarType
+prioritize(at::ScalarType current, const Tensor& nextArg) {
+  if (nextArg.is_cuda() && nextArg.is_floating_point()) {
+    auto next = nextArg.scalar_type();
+    // For promotion purposes, prioritize double, then float, then half.
+    if (current == at::kDouble || next == at::kDouble) {
+      return at::kDouble;
+    } else if (current == at::kFloat || next == at::kFloat) {
+      return at::kFloat;
+    } else if (current == at::kHalf && next == at::kHalf) {
+      return at::kHalf;
+    } else {
+      AT_ERROR("Unexpected floating ScalarType in at::autocast::prioritize");
+      return current;
+    }
+  } else {
+    return current;
+  }
+}
+
+// Overload to catch TensorList args (for e.g. cat, stack).
+// Reuses the overload above to process each Tensor in the list.
+inline at::ScalarType
+prioritize(at::ScalarType current, const TensorList& list) {
+  for (const auto& tensor : list) {
+    current = prioritize(current, tensor);
+  }
+  return current;
+}
+
+// Template to catch non-Tensor args (no-op that returns current best guess)
+template<typename T> inline at::ScalarType
+prioritize(at::ScalarType current, T nextArg) {
+  return current;
+}
+
+// Overload for the tail case.
+inline at::ScalarType
+promote_type(at::ScalarType current) {
+  return current;
+}
+
+// Unpack args and pick the widest floating-point dtype among Tensors to use for promotion.
+// Non-Tensor arguments are ignored.
+template<typename Arg0, typename... Args> inline at::ScalarType
+promote_type(at::ScalarType current, Arg0 arg0, Args... args) {
+  auto new_current = prioritize(current, arg0);
+  return promote_type(new_current, args...);
+}
+
+/****************************************************
+Logic to apply cached casting to any Tensor argument.
+****************************************************/
+
+// Overload to catch Tensor args
+inline Tensor
+cached_cast(at::ScalarType to_type, const Tensor& arg) {
+  if (arg.is_cuda() && arg.is_floating_point() && arg.scalar_type() != to_type) {
+    // Heuristic:  Do what Apex does, and cache FP16 casts of FP32 model weights (leaves).
+    // See cached_casts declaration above for detailed strategy.
+    bool can_try_cache = (to_type == at::kHalf && arg.scalar_type() == at::kFloat && arg.requires_grad() && arg.is_leaf());
+    if (can_try_cache) {
+      auto it = cached_casts.find(arg.unsafeGetTensorImpl());
```

Comment:
Need a lock, right?

Reply:
cast cache is thread_local.  That's another decision we may wish to discuss:  should the cast cache be thread_local or shared across threads and mutexed?  Or (dare I say it) should it not exist in the first PR?  It's helpful primarily for cell-based RNNs to avoid casting inputs over and over.

Reply:
I definitely want to have a broader discussion with you, @ngimel, and @mruberry about the philosophy of the cast cache.  I'd appreciate if we can talk about this on slack whenever is convenient.

Reply:
Hmm, thread local seems fine to start, anycase.